### PR TITLE
Fix bot throws an error due to the new version has conflict with the restored conversation state

### LINKED ISSUE #5007: Bot throws error when new version has been deployed
In existing conversations with bots, when a new version of the bot is deployed an error is thrown when the user continues the conversation. 

image

When this situation happens the bot should have a failback, as uploading a new version is part of continuous improvement and conversations can be long lasting, like in Teams for example.

### PR DESCRIPTION
**Pull request by cosmicshuai**
Closes #5007 

During the continous deployment of bots, while bot assets changed and the bot resume the state from persistent  storage, the bot may throws error. It is caused by the dialogId may no longer exsisted in the bot assets. 
In this PR, if such error occured, the current dialog will be replaced by the root dialog.

### CODE REVIEW

**Comment by LeeParrishMSFT** [comment]
@EricDahlvang Can you take a look at this one? Thanks!

**Comment by EricDahlvang** [comment]
@LeeParrishMSFT  this is a draft, and does not look ready for review.

Also, i'm not seeing how these changes would resolve the issue: https://github.com/microsoft/botbuilder-dotnet/issues/5007   

The changes in this PR currently do not look like they will have any affect, since DialogContext.FindDialog already calls the parent recursively:

```cs
                if (this.Dialogs != null)
                {
                    var dialog = this.Dialogs.Find(dialogId);
                    if (dialog != null)
                    {
                        return dialog;
                    }
                }

                if (this.Parent != null)
                {
                    return this.Parent.FindDialog(dialogId);
                }
```

The problem is, the ids are re-written as the dialog stack is changed.  Until the dialog ids are not re-written, or there is a mapping/coordination between previous and new ids, as well as dialog versioning stored with the state: this will continue to be a problem.

**Comment by cosmicshuai** [comment]
Hi @EricDahlvang,
I completely agree that the only way to continue the dialog is keep the IDs unchanged or we have a old-new IDs mapping.
So the current solution only tried to resume the dialog in some point. The difference of the findDialog is, 
The existing code:
```cs
                if (this.Dialogs != null)
                {
                    var dialog = this.Dialogs.Find(dialogId);
                    if (dialog != null)
                    {
                        return dialog;
                    }
                }

                if (this.Parent != null)
                {
                    return this.Parent.FindDialog(dialogId);
                }
```
Only tries to recursively find the dialog of the ActiveID of current dialogContext. 

The code I added:
```cs

                    if (dialog == null)
                    {	
                        var dc = this;
                        while (dialog == null && dc != null)
                        {
                            dc = dc.Parent;
                            dialog = dc.FindDialog(dc.ActiveDialog.Id);
                        }

                        if (dialog == null)
                        {
                            throw new InvalidOperationException($"Failed to continue dialog. A dialog with id {this.ActiveDialog.Id} could not be found.");
                        }
                        else
                        {
                            return await dialog.BeginDialogAsync(dc, cancellationToken).ConfigureAwait(false);
                        }
```

once the previous findDialog returns nothing, it will recursively find a valid activeDialog of current dialogContext's ancestor dialogContext. Then restart the dialog from there. 

This is only a draft solution. The default behavior fallback of no dialog found need to be discussed further. I will keep talking to Dong and Tom to find out a better solution.

**Comment by carlosscastro** [comment]
Thanks for debugging @cosmicshuai! I'd love to discuss this with you folks and bring in @tomlm and @Stevenic. 

Let's consider a deployment and whether assets or not: 

1- Assets did not change at all: Then all the IDs are recomputed to the same values and we guarantee that the conversation continues smoothly
2- Assets changed: This is what we need to discuss. In the case you are fixing, the dialog that got changed was the dialog that was on the top of the stack when the conversation was persisted. In that case, the trick of finding the dialog up the parents stack could help identify the active dialog. However the extent to which dialogs could change is endless, including swapping ID's and absolutely changing the structure of the dialog graph, etc. It's not obvious 1) how to determine whther the change was large enough that the data in storage is not relevant to the new assets and 2) what to do if we decide that the dialog structure changed in such a way that we cannot use part of the state.

Let's have a discussion on what is the right approach here and take it from there.

**Comment by cosmicshuai** [comment]
Update: After some discussion, the current solution is to restart the whole conversation once the dialog assets changed during the CI/CD. Still need some debug to ensure everything works correctly.

**Comment by cosmicshuai** [comment]
@carlosscastro  
Could you help review this PR?

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs [hunk 1]
```diff
@@ -354,7 +354,18 @@ public async Task<DialogTurnResult> EndDialogAsync(object result = null, Cancell
...
                     {
-                        throw new InvalidOperationException($"DialogContext.EndDialogAsync(): Can't resume previous dialog. A dialog with an id of '{ActiveDialog.Id}' wasn't found.");
+                        var rootDialog = this.DialogManager.RootDialog;
+
+                        try
+                        {
+                            return await this.ReplaceDialogAsync(rootDialog.Id, cancellationToken: cancellationToken).ConfigureAwait(false);
+                        }
+#pragma warning disable CS0168 // Variable is declared but never used
+                        catch (Exception err)
```
Comment (mrivera-ms) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs:364]: If you are not using the exception, don't need declare a variable for it. Please remove the pragma.

**Comment by mrivera-ms** [comment]
@cosmicshuai, looks like this fix is still under discussion.  Should we convert this back to draft until it is decided?

**Comment by cosmicshuai** [comment]
@mrivera-ms Sure, please convert it to draft.
I think I need to add a unit test that once the bot use conversation state and the bot assets changed. It won't throw this error again.

**Comment by cosmicshuai** [comment]
Trying to add a e2e test for this fix.

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs (hunk 1 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs]: Thanks, removed!

**Comment by mrivera-ms** [comment]
@cosmicshuai, @boydc2014, are we on track for R13? If not, we need to push the issue to the next milestone. Thank you.

**Comment by cosmicshuai** [comment]
Here is a E2E test for this fix. 

1. Add CosmosDB to TestBot Json
image
image

2. Add a bot use cosmosdb to store conversation state.

3. change the bot assets then resume the dialog

I recorded a video to demo this change:

https://user-images.githubusercontent.com/17074777/112590175-59763380-8e3d-11eb-8161-0ac8cf76eca9.mp4

**Review by boydc2014 — APPROVED** [review]

**Comment by boydc2014** [comment]
> @cosmicshuai, @boydc2014, are we on track for R13? If not, we need to push the issue to the next milestone. Thank you.

Yes, the code is ready, aligned with Shuai and Carlos's conversation. Shuai just updated some more proof, and i just approved that.

**Comment by cosmicshuai** [comment]
@mrivera-ms Hi Monica, I think this PR is ready for merge. Could you take a look?

**Comment by mrivera-ms** [comment]
@carlosscastro, could you please take a look at this PR? @cosmicshuai has updated the changes.

**Comment by drmanhatin** [comment]
Please merge this pr, so I don't have to write a workaround.

**Comment by gabog** [comment]
Closing this one for now, the proposed solution doesn't address the problem and could generate others.

**Pull request closed without merging**